File: course/section_4_stack/code/03-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack: list[str] = []
        open_to_close = {")": "(", "]": "[", "}": "{"}
        for char in s:
            if char in open_to_close:
                if stack and stack[-1] == open_to_close[char]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char)
        return not stack

File: course/section_4_stack/code/03-valid-parentheses.php
<?php
class Solution {
    function isValid($s) {
        $stack = [];
        for($i = 0 ; $i < strlen($s); $i++){
            $bracket = $s[$i];
            if($bracket === '[' || $bracket === '(' || $bracket === '{'){
                $stack[] = $bracket ;
            } else {
                if(empty($stack)){
                    return false ;
                } elseif(end($stack) === '(' and $bracket === ')'){
                    array_pop($stack);
                } elseif(end($stack) === '[' and $bracket === ']') {
                    array_pop($stack);
                } elseif(end($stack) === '{' and $bracket === '}') {
                    array_pop($stack);
                } else {
                    return false ;
                }
            }
        }
        return empty($stack);
    }
}